src/backend/executor/functions.c

```diff
@@ -944,7 +944,11 @@ postquel_start(execution_state *es, SQLFunctionCachePtr fcache)
 		else
 		{
 			/* Otherwise, we do not record information about internal queries. */
-			es->qd->gpmon_pkt = NULL;
```

Comment:
What about similar [place ](https://github.com/greenplum-db/gpdb/blob/6109de10d3d60563f777f41ed7cb1a76e1c52353/src/backend/executor/spi.c#L2318)for SPI? But, as far I Understand, plpgsql functions that use SPI, are executed in the separate `_SPI_current->execCxt` memory context, that will be reset each time.

Reply:
I have created function
```
CREATE TABLE t(i int, j int);

CREATE OR REPLACE FUNCTION somefunc(integer) RETURNS integer as $$
DECLARE
 s integer;
BEGIN
 FOR k IN 1..1000000 LOOP
  SELECT SUM(i) INTO s FROM t WHERE i < $1;
 END LOOP;
 RETURN s;
END;
$$ LANGUAGE plpgsql;
```
and I have called it. CreateQueryDesc() and FreeQueryDesc() is called for each SELECT, but memory does not leak, because gpmon_pkt is released after each SELECT in _SPI_end_call(). memleak does not show stack with CreateQueryDesc(). There is no need to free memory at this code block where pointer becomes NULL.

Reply:
IMHO explicit free in this case is not so bad

Reply:
Let's sync these code parts as far as _SPI_end_call resets memory context depending on an argument. This argument is always true currently, but...